vql/linux/audit.go

```diff
@@ -4,90 +4,688 @@ package linux
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"time"
-        "strings"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
 
 	"github.com/Velocidex/ordereddict"
 	"github.com/elastic/go-libaudit"
 	"github.com/elastic/go-libaudit/aucoalesce"
 	"github.com/elastic/go-libaudit/auparse"
-        "github.com/elastic/go-libaudit/rule"
+	auditrule "github.com/elastic/go-libaudit/rule"
 	"github.com/elastic/go-libaudit/rule/flags"
 	"www.velocidex.com/golang/velociraptor/acls"
+	"www.velocidex.com/golang/velociraptor/artifacts"
+	config_proto "www.velocidex.com/golang/velociraptor/config/proto"
+	"www.velocidex.com/golang/velociraptor/logging"
 	vql_subsystem "www.velocidex.com/golang/velociraptor/vql"
 	"www.velocidex.com/golang/vfilter"
-        "www.velocidex.com/golang/vfilter/arg_parser"
+	"www.velocidex.com/golang/vfilter/arg_parser"
 )
 
-var auditrules [][]byte
+var (
+	gLock sync.Mutex
+	gService *AuditWatcherService
+	// Timeout for batching audit configuration events
+	gBatchTimeout = 1000 * time.Millisecond
+	debugGoRoutines = false
+)
+
+var gBannedRules = []string{
+	"-d task,never",
+}
+
+type AuditWatcherRule struct {
+	wfRule	auditrule.WireFormat
+	rule	string
+}
+
+type AuditWatcher struct {
+	id		int
+	ctx		context.Context
+	eventChannel	chan vfilter.Row
+	// Handles logging for each invocation.  The logger itself is
+	// threadsafe and we don't need it to be from the same thread as the initial
+	// invocation.
+	scope		vfilter.Scope
+	rules		map[string]*AuditWatcherRule
+	auditService	*AuditWatcherService
+}
+
+func parseRule(rule string) (*AuditWatcherRule, error) {
+	r, err := flags.Parse(rule)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse rule `%s': %v", rule, err)
+	}
+
+	wfRule, err := auditrule.Build(r)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to build rule `%s': %v", rule, err)
+	}
+
+	normalizedRule, err := auditrule.ToCommandLine(wfRule, true)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to normalize rule `%s': %v", rule, err)
+	}
+
+	watcherRule := &AuditWatcherRule{
+		wfRule:	wfRule,
+		rule:	normalizedRule,
+	}
+
+	return watcherRule, nil
+}
+
+func (self *AuditWatcher) addRules(rules []string) error {
+	for _, line := range rules {
+		watcherRule, err := parseRule(line)
+		if err != nil {
+			return err
+		}
+
+		_, ok := self.rules[watcherRule.rule]
+		if ok {
+			return fmt.Errorf("Cannot add duplicate rule `%s' in single query", line)
+		}
+
```

Comment:
nit: this could be written as 

```
if _, ok := self.rules[watcherRule.rule]; ok { 
       return fmt.Errorf("Cannot add duplicate rule `%s' in single query", line)
}
```

Reply:
It could, but I really don't like that pattern.